File: protocol/impls/go/state.go
package main

// Fold state. All collections are keyed deterministically; digest.go emits them
// in the conformance-pinned sort order (never by Go map iteration).
// Names-only: names + commits + muts (no votes / posts / decorations).

type NameRow struct {
	owner [20]byte
	st    int // ST_OWNED / ST_LISTED / ST_OFFERED / ST_RESERVED

	leaseExpiry int64

	// market fields — always digested; zeroed when not active for the current st.
	seller        [20]byte
	sellerType    byte
	price         uint64
	offerExpiry   int64
	buyer         [20]byte // directed-offer buyer OR open-reserve reserver
	burnLeg       uint64
	payLeg        uint64
	reserveExpiry int64
}

// resetMarket physically zeroes all market fields (used on close/revert and OWNED).
func (r *NameRow) resetMarket() {
	r.seller = [20]byte{}
	r.sellerType = 0
	r.price = 0
	r.offerExpiry = 0
	r.buyer = [20]byte{}
	r.burnLeg = 0
	r.payLeg = 0
	r.reserveExpiry = 0
}

type CommitRow struct {
	commitment [32]byte
	height     int64
	txIndex    uint32
	commitTime int64
}

type FoldState struct {
	names   map[string]*NameRow // key = raw name bytes
	commits []CommitRow
	muts    map[[20]byte]int64

	curHeight int64

	// per-block claim displacement scratch (reset each block; NOT digested).
	scratch map[string]claimScratch
}

type claimScratch struct {
	ch    int64
	ti    uint32
	owner [20]byte
}

func newState() *FoldState {
	return &FoldState{
		names: map[string]*NameRow{},
		muts:  map[[20]byte]int64{},
	}
}

func (s *FoldState) clear() {
	*s = *newState()
}

func (s *FoldState) bumpMut(owner [20]byte, h int64) {
	if cur, ok := s.muts[owner]; !ok || h > cur {
		s.muts[owner] = h
	}
}

// makeSyntheticTxid: txid = u64_le(height) ‖ u32_le(txindex) ‖ 20 zero bytes (conformance §3).
func makeSyntheticTxid(height int64, txindex uint32) [32]byte {
	var t [32]byte
	h := uint64(height)
	for i := 0; i < 8; i++ {
		t[i] = byte(h >> (8 * uint(i)))
	}
	t[8] = byte(txindex)
	t[9] = byte(txindex >> 8)
	t[10] = byte(txindex >> 16)
	t[11] = byte(txindex >> 24)
	return t
}

File: protocol/impls/java/State.java
import java.math.BigInteger;
import java.util.*;

// Live indexer state — the only thing an indexer MUST keep (§3.9), and exactly
// what the canonical digest (§4) serializes. Names keyed by their raw ASCII bytes
// (charset [a-z0-9-], already lowercase = canonical key, §3.1).
final class State {
    static final byte[] ZERO20 = new byte[20];
    static final byte[] ZERO32 = new byte[32];

    static final class NameRow {
        byte[] owner = ZERO20;
        int ownerType;                 // NOT digested (cosmetic, §3.9) — tracked only
        int st = Const.OWNED;          // OWNED/LISTED/OFFERED/RESERVED
        long leaseExpiry;
        // listing / offer
        byte[] seller = ZERO20;
        int sellerType;                // digested
        BigInteger price = BigInteger.ZERO;
        long offerExpiry;
        // reservation (reserved) OR named buyer (offered)
        byte[] buyer = ZERO20;
        BigInteger burnLeg = BigInteger.ZERO;
        BigInteger payLeg = BigInteger.ZERO;
        long reserveExpiry;
    }

    static final class Commit {
        byte[] commitment;             // 32
        long commitHeight;
        long txIndex;                  // u32 — the COMMIT's position in its block (§3.2 tuple)
        long commitTime;               // MTP at commit (COMMIT_EXPIRY window)
    }

    static final class Vote {
        byte[] target;                 // 32
        long vout;
        BigInteger score = BigInteger.ZERO; // signed i128 accumulator
    }

    static final class Decor {
        byte[] txid;                   // 32 (synthetic post id)
        long vout;
        byte[] rec;                    // one TLV record, verbatim
        int seq;                       // insertion order (stable tiebreak within a post)
    }

    final Map<String, NameRow> names = new HashMap<>();
    final List<Commit> commits = new ArrayList<>();
    final Map<String, Vote> votes = new LinkedHashMap<>();   // key = hex(target)+":"+vout
    final Map<String, Long> muts = new HashMap<>();           // key = hex(owner) -> last_set_mutation_height
    final List<Decor> decors = new ArrayList<>();
    boolean overflow = false;
    private int decorSeq = 0;

    // per-block claim scratch (NOT digested): name -> {commit_height, owner, commit tx_index}
    static final class ClaimMark { long commitHeight; byte[] owner; long commitTxIndex; }
    final Map<String, ClaimMark> claimScratch = new HashMap<>();

    void clear() {
        names.clear(); commits.clear(); votes.clear(); muts.clear(); decors.clear();
        claimScratch.clear(); overflow = false; decorSeq = 0;
    }

    int nextDecorSeq() { return decorSeq++; }

    // ---- helpers -----------------------------------------------------------

    static String nameKey(byte[] n) { return new String(n, java.nio.charset.StandardCharsets.US_ASCII); }

    void bumpMut(byte[] owner, long height) { muts.put(Hex.enc(owner), height); }

    long lastMut(byte[] owner) { Long h = muts.get(Hex.enc(owner)); return h == null ? Long.MIN_VALUE : h; }

    // owned-set of an owner, names lexicographically (raw-byte) ascending — the
    // bitmap ordering (§3.5). "owned" = the name exists in the set with this owner,
    // regardless of listed/offered (a listed/offered name keeps its bitmap position).
    List<String> ownedSetSorted(byte[] owner) {
        String oh = Hex.enc(owner);
        List<String> r = new ArrayList<>();
        for (Map.Entry<String, NameRow> e : names.entrySet())
            if (Hex.enc(e.getValue().owner).equals(oh)) r.add(e.getKey());
        r.sort(State::cmpNameKey);
        return r;
    }

    // raw-byte lexicographic comparison of two ASCII name keys.
    static int cmpNameKey(String a, String b) {
        byte[] x = a.getBytes(java.nio.charset.StandardCharsets.US_ASCII);
        byte[] y = b.getBytes(java.nio.charset.StandardCharsets.US_ASCII);
        return cmpBytes(x, y);
    }

    static int cmpBytes(byte[] x, byte[] y) {
        int n = Math.min(x.length, y.length);
        for (int i = 0; i < n; i++) {
            int d = (x[i] & 0xFF) - (y[i] & 0xFF);
            if (d != 0) return d;
        }
        return x.length - y.length;
    }
}
